libraries/chain/db_maint.cpp

```diff
@@ -198,105 +198,112 @@ void database::pay_workers( share_type& budget )
+            }
+
+            // total_votes is 64 bits. Subtract the number of leading low bits from 64 to get the number of useful bits,
+            // then I want to keep the most significant 16 bits of what's left.
+            int8_t bits_to_drop = std::max(int(boost::multiprecision::detail::find_msb(total_votes)) - 15, 0);
+            for( const auto& weight : weights )
+            {
+               // Ensure that everyone has at least one vote. Zero weights aren't allowed.
+               uint16_t votes = std::max((weight.second >> bits_to_drop), uint64_t(1) );
+               a.active.account_auths[weight.first] += votes;
+               a.active.weight_threshold += votes;
+            }
 
-         // total_votes is 64 bits. Subtract the number of leading low bits from 64 to get the number of useful bits,
-         // then I want to keep the most significant 16 bits of what's left.
-         int8_t bits_to_drop = std::max(int(boost::multiprecision::detail::find_msb(total_votes)) - 15, 0);
-         for( const auto& weight : weights )
+            a.active.weight_threshold /= 2;
+            a.active.weight_threshold += 1;
+         }
+         else
          {
-            // Ensure that everyone has at least one vote. Zero weights aren't allowed.
-            uint16_t votes = std::max((weight.second >> bits_to_drop), uint64_t(1) );
-            a.active.account_auths[weight.first] += votes;
-            a.active.weight_threshold += votes;
+            vote_counter vc;
+            for( const witness_object& wit : wits )
+               vc.add( wit.witness_account, _vote_tally_buffer[wit.vote_id] );
+            vc.finish( a.active );
          }
+      } );
 
-         a.active.weight_threshold /= 2;
-         a.active.weight_threshold += 1;
-      }
-      else
+      modify( gpo, [&wits]( global_property_object& gp )
       {
-         vote_counter vc;
-         for( const witness_object& wit : wits )
-            vc.add( wit.witness_account, _vote_tally_buffer[wit.vote_id] );
-         vc.finish( a.active );
-      }
-   } );
-
-   modify( gpo, [&wits]( global_property_object& gp )
-   {
-      gp.active_witnesses.clear();
-      gp.active_witnesses.reserve(wits.size());
-      std::transform(wits.begin(), wits.end(),
-                     std::inserter(gp.active_witnesses, gp.active_witnesses.end()),
-                     [](const witness_object& w) {
-         return w.id;
+         gp.active_witnesses.clear();
+         gp.active_witnesses.reserve(wits.size());
+         std::transform(wits.begin(), wits.end(),
+                        std::inserter(gp.active_witnesses, gp.active_witnesses.end()),
+                        [](const witness_object& w) {
+            return w.id;
+         });
       });
-   });
+   }
+
+   _active_witnesses.clear();
+   _active_witnesses.reserve( gpo.active_witnesses.size() );
+   for( const witness_id_type& wid : gpo.active_witnesses )
+      _active_witnesses.push_back( &(wid(*this)) );
```

Comment:
IMHO this change is too much although only changed spaces. Why not add these 4 lines to a new function and call it when needed? Update: the parameter is always false, so why need to change this function at all?

Reply:
I originally put it in a new function, but then did it this way as I couldn't figure out a decent name for the function. "update_active_witnesses" is already taken ;)

update_active_witnesses is now called during maintenance, and during database opening. Hence, the default parameter.

I will review, and see if I can't come up with a decent name and split it out.